edit_questionpy_form.php

```diff
@@ -70,6 +71,81 @@ public function __construct(string $submiturl, object $question, object $categor
         parent::__construct($submiturl, $question, $category, $contexts, $formeditable);
     }
 
+
+    /**
+     * Returns the question state of the stored question if possible.
+     *
+     * If the question is being edited and the package (version) was changed, no question state is returned.
+     *
+     * @param string $packagehash
+     * @return string|null
+     */
+    private function get_question_state(string $packagehash): ?string {
+        if (isset($this->question->qpy_id) && $this->question->qpy_package_hash === $packagehash) {
+            return $this->question->qpy_state;
+        }
+        return null;
+    }
+
+    /**
+     * Adds alert to the form to warn that the selected package is not usable even though it is available through the package
+     * search.
+     *
+     * @param MoodleQuickForm $mform
+     * @param string $hash
+     * @return void
+     * @throws moodle_exception
+     */
+    private function definition_package_not_usable(MoodleQuickForm $mform, string $hash): void {
+        global $OUTPUT;
+
+        // The current package is in use if the current question is saved and its package hash matches the hash of the selected
+        // package.
+        $packageinuse = isset($this->question->qpy_id) && $this->question->qpy_package_hash == $hash;
+
+        // Create a group which contains the alert - the group is used to simplify the styling.
+        $group[] = $mform->createElement('html', $OUTPUT->render_from_template(
+            'qtype_questionpy/package/package_not_usable',
+            ['inuse' => $packageinuse],
+        ));
+        $mform->addGroup($group, '');
+
+        $mform->addElement('hidden', 'qpy_package_invalid', true);
+        $mform->setType('qpy_package_invalid', PARAM_BOOL);
+    }
+
+    /**
+     * Adds alert to the form to warn that the selected package is not available anymore.
+     *
+     * @param MoodleQuickForm $mform
+     * @param string $hash
+     * @return void
+     * @throws moodle_exception
+     */
+    private function definition_package_not_available(MoodleQuickForm $mform, string $hash): void {
+        global $DB, $OUTPUT;
+
+        // The current package is in use if the current question is saved and its package hash matches the hash of the selected
+        // package.
+        $packageinuse = isset($this->question->qpy_id) && $this->question->qpy_package_hash == $hash;
```

Comment:
Kann der `false`-Fall überhaupt bei "normaler" Verwendung auftreten? Wenn das Paket in der DB fehlt, wird es doch nicht in der Suche angezeigt, d.h. eine Frage, die das Paket referenziert, muss in der Vergangenheit erstellt worden sein. Ich denke nicht, dass wir die Fehlermeldung explizit für den Fall anpassen müssen, dass der DB-Eintrag während der Frage-Erstellung gelöscht wird.